### ISSUE TITLE
about criterion

### ISSUE THREAD

**Issue by ForeverPs** [zycdev/L2R2#1]
What is the format of y_true, the rank order?
Does smaller number represents more relevant item?

**Comment by zycdev** [comment]
`y_true` is the relevance level label, the bigger the more relevant.

**Comment by ForeverPs** [comment]
Many thanks for your great work!
![image](https://user-images.githubusercontent.com/47919727/150263001-02fc8999-ccfd-4515-9346-297b0966c329.png)
In my experiments, I give bigger target number to the item with greater relevance. As in the image, the first tensor is my target input to the criterion, and the second tensor is the groundtruth relevance, however, the loss of my net can not decrease normally. Looking forward to your reply.

**Comment by zycdev** [comment]
I'm wondering if you're saying the opposite. From the picture, the first tensor is more like the target relevance label, and the second is more like the predicted logits. If there is no problem with the model, all the loss functions in [losses.py](https://github.com/zycdev/L2R2/blob/master/losses.py) can backward normally.

**Comment by ForeverPs** [comment]
Many Thanks，things go well ! In ListMLE loss, I wonder why you minus the max predicted logit, it seems that this operation does not change the probability. Maybe, this is for calculation stability? exp(max predict logit) is too big?
![image](https://user-images.githubusercontent.com/47919727/150514002-f4d3cb7f-ca75-4ed2-8fb6-0928923a1dd0.png)
In losses.py list_mle()

**Comment by lxianl455** [comment]
@ForeverPs 兄弟，你问题解决了吗？我用这些loss排序，loss也不下降。训练很多个epoch，loss都只降一点点，大部分时候十分震荡，而且最后效果都很不理想。

**Comment by lxianl455** [comment]
我一样用的是排名作为label

**Comment by ForeverPs** [comment]
> 我一样用的是排名作为label

是的，排序损失相当于是一个粒度非常细的分类问题，本身就是比较困难的。我后来是先做粗粒度的分类，然后在每个类别里面做了ranking，会好一些

**Comment by lxianl455** [comment]
> > 我一样用的是排名作为label
> 
> 是的，排序损失相当于是一个粒度非常细的分类问题，本身就是比较困难的。我后来是先做粗粒度的分类，然后在每个类别里面做了ranking，会好一些

谢谢回复，那我再摸索一下

**Comment by ForeverPs** [comment]
> 

可以尝试一下Margin loss，也是一种ranking loss，会比sigmoid类型的要稳定一些。margin based ranking loss

**Comment by lxianl455** [comment]
> > 
> 
> 可以尝试一下Margin loss，也是一种ranking loss，会比sigmoid类型的要稳定一些。margin based ranking loss

是的，我一开始尝试的是Margin rank loss，效果还算可以，然后直观上感觉没有用到list的约束，感觉list wise会比pair wise的会好，所以想尝试一下list wise。对于你的评论中”sigmoid类型的要稳定一些“，应该如何理解？我确实观察到list_net 和 lise_mle 都经过了一个归一化的过程，而如果用margin rank loss则不需要归一化的过程。

**Comment by ForeverPs** [comment]
> > > 
> > 
> > 
> > 可以尝试一下Margin loss，也是一种ranking loss，会比sigmoid类型的要稳定一些。margin based ranking loss
> 
> 是的，我一开始尝试的是Margin rank loss，效果还算可以，然后直观上感觉没有用到list的约束，感觉list wise会比pair wise的会好，所以想尝试一下list wise。对于你的评论中”sigmoid类型的要稳定一些“，应该如何理解？我确实观察到list_net 和 lise_mle 都经过了一个归一化的过程，而如果用margin rank loss则不需要归一化的过程。

sigmoid会把logits映射到0-1之间，但是sigmoid函数的感受范围很小，在-5或者是5以外，就已经进入饱和区了，都是0或者1，这样是不利于作排序的，margin loss是直接对logits作损失，总是可以得到顺序。并且在我自己的实验中，sigmoid表现不是很好

**Comment by lxianl455** [comment]
确实，实践下来感觉你这个总结得很有道理，谢谢兄弟指导

**Comment by lxianl455** [comment]
> > > > 
> > > 
> > > 
> > > 可以尝试一下Margin loss，也是一种ranking loss，会比sigmoid类型的要稳定一些。margin based ranking loss
> > 
> > 
> > 是的，我一开始尝试的是Margin rank loss，效果还算可以，然后直观上感觉没有用到list的约束，感觉list wise会比pair wise的会好，所以想尝试一下list wise。对于你的评论中”sigmoid类型的要稳定一些“，应该如何理解？我确实观察到list_net 和 lise_mle 都经过了一个归一化的过程，而如果用margin rank loss则不需要归一化的过程。
> 
> sigmoid会把logits映射到0-1之间，但是sigmoid函数的感受范围很小，在-5或者是5以外，就已经进入饱和区了，都是0或者1，这样是不利于作排序的，margin loss是直接对logits作损失，总是可以得到顺序。并且在我自己的实验中，sigmoid表现不是很好

确实，实践下来感觉你这个总结得很有道理，谢谢兄弟指导
